clang/test/BoundsSafety/AST/objc.m

```diff
@@ -0,0 +1,154 @@
+// RUN: %clang_cc1 -fbounds-safety -x objective-c -fbounds-attributes-objc-experimental -Wno-deprecated-declarations -Wno-return-type -Wno-objc-root-class -ast-dump %s 2>&1 | FileCheck %s
+// RUN: %clang_cc1 -fexperimental-bounds-safety-attributes -x objective-c -Wno-deprecated-declarations -Wno-return-type -Wno-objc-root-class -ast-dump %s 2>&1 | FileCheck %s
+
+#include <ptrcheck.h>
+
+@protocol CountedByProtocol
+ - (void)simpleParam:(int)len :(int * __counted_by(len))p;
+ - (void)reverseParam:(int * __counted_by(len))p :(int)len;
+ - (void)nestedParam:(int*)len :(int * __counted_by(*len))p;
+
+ - (int * __counted_by(len))simpleRet:(int)len;
+ - (int * __counted_by(*len))nestedRet:(int*)len;
+
+ - (void)cParam:(int)len, int * __counted_by(len) p;
+ - (void)reverseCParam:(int * __counted_by(len))p, int len;
+@end
+
+// CHECK-LABEL: -ObjCProtocolDecl {{.*}} CountedByProtocol
+// CHECK-NEXT:   |-ObjCMethodDecl {{.*}} - simpleParam:: 'void'
+// CHECK-NEXT:   | |-ParmVarDecl {{.*}} len 'int'
+// CHECK-NOT: IsDeref
+// CHECK-NEXT:   | | `-DependerDeclsAttr
+// CHECK-NEXT:   | `-ParmVarDecl {{.*}} p 'int *{{(__single)?}} __counted_by(len)':'int *{{(__single)?}}'
```

Comment:
We don't make thin pointers  `__single` in attribute-only modes, so the expected output should be different between the two modes.